Switch to Make.

bentley opened the pull request:
Packagers tend not to be fans of scons. [The Gentoo wiki lists some reasons](https://wiki.gentoo.org/wiki/SCons); at OpenBSD we share those concerns and have a few of our own.

Make is simple and widely portable. Here's a Makefile (compatible with at least BSD Make and GNU Make ≥ 4.0) that builds libgambatte and gambatte_sdl.

On my laptop, build time goes down from 53 seconds (scons) to 20 seconds (make).

sinamas committed:
rgb32ToRgb16: save 1 instruction in inner-loop on ARM (gcc).
gcc is bad at folding shifts into ands (and probably other ops) with simple constants (it prefers using immediates, saving a register and reducing loop header).
help it fold shifts into the ors (non-constant, so gcc cannot pick the immediate version) instead.


git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@330 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
ppu: shave off 5 instructions in unrolled tile-draw inner loop on ARM (gcc)
help gcc fuse the shifts with the loads since it doesn't manage to fuse them with the ands.


git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@331 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
delete man pages


git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@332 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
m0 irq inhibits lycirq trigger.


git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@333 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
missing ds ly reg 99 read tests. longer 99 period when ds.


git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@334 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
plug gaping hole in lyc99 flag read behavior.


git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@335 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
better lyc trigger inhibition hypothesis with evidence.
unify and simplify lyc cmp code.


git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@336 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
whitespace


git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@337 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
missing ds tests.


git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@338 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
fix test name and cgb flag.


git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@339 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
delete empty dir


git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@340 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
add saveSavedata method. based on patch from Robert Broglia.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@341 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
saveState: return success. based on patch from Robert Broglia.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@342 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
loadState: return success. based on patch from Robert Broglia.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@343 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
define OF if not defined.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@344 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
return more informative load result.

don't print errors in libgambatte.
use error code string in ports for more informative error message.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@345 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
add game pak info method. do not print in libgambatte.

do print rom title and game pak info in qt/sdl ports.
add header checksum info.
add [bad] to qt window title if header checksum is bad.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@346 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
print 'c'/'d' (cgb/dmg) for each test.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@347 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
do not include cstdio. not needed anymore.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@348 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
update and reformulate. s/-end/end/.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@349 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
refactor.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@350 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
s/getDesc/desc/

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@351 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
s/getStr/str. s/getChar/character.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@352 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
break line.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@353 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
add cli option to show keyboard controls.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@354 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
replace *const with [].

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@355 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
const

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@356 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
scope

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@357 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
get rid of cast, sizeof(type).

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@358 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
s/aamas@stud.ntnu.no/sinamas@users.sourceforge.net/

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@359 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
change default audio buffer latency to 68 ms.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@360 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
testrunner: link to libm. neded on OpenBSD

Contributed by Anthony J. Bentley (anthonyjbentley at users.sf.net)

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@361 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
ossengine: drop "sys/" from soundcard.h include-path if OpenBSD

suggested by Anthony J. Bentley ( anthonyjbentley ).

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@362 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
qt/framework: define USBHID_NEW and USBHID_UCR_DATA on OpenBSD

may fix joystick build issues on OpenBSD.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@363 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
libgambatte: fix psg ch3 regression introduced in r165

delta optimizations broke ch3 dac off output level.

git-svn-id: https://gambatte.svn.sourceforge.net/svnroot/gambatte@364 9dfb2916-2d38-0410-aef4-c5fe6c9ffc24

sinamas committed:
common: add defined_ptr/poor mans boost::checked_delete

sinamas committed:
utilize defined_ptr

and prune some cruft.

sinamas committed:
common: add poor man's unique_ptr

all in the name of not depending on tr1, c++11 or boost (stoneage c++98
ftw).

a marginal step up from std::auto_ptr but better than nothing imo.

sinamas committed:
libgambatte: replace use of std::auto_ptr with tranfer/scoped_ptr

sinamas committed:
gambatte_sdl: replace use of std::auto_ptr with scoped_ptr

sinamas committed:
atomicvar: missing includes hidden by include order

sinamas committed:
gambatte_qt: replace use of std::auto_ptr with scoped/transfer_ptr

sinamas committed:
mediaworker: remedy inexplicable lack of QMutexLocker

sinamas committed:
mediaworker: nonfunctional/readability. ws mostly.

sinamas committed:
audioengine: remedy inexplicable lack of QMutexLocker

sinamas committed:
libgambatte: c++11 narrowing conversion warnings

sinamas committed:
qglblitter: shut up warning

sinamas committed:
gambatte_qt: c++11 narrowing conversion warnings

sinamas committed:
gambatte_sdl: c++11 narrowing conversin warnings

sinamas committed:
mediaworker: prefer deque::assign to custom loop

sinamas committed:
statesaver: more conventional/ideomatic style

sinamas committed:
bitmap_font: std::reverse exists. use it.

sinamas committed:
libgambatte: more ideomatic use of sizeof

sinamas committed:
scalebuffer: more ideomatic use of sizeof

sinamas committed:
common: more ideomatic use of sizeof

sinamas committed:
openalengine: delete dead code

sinamas committed:
gambatte_qt: more ideomatic use of sizeof

sinamas committed:
add .gitignore

sinamas committed:
usec: delete superflous noise

sinamas committed:
gambatte_qt: excessive parentheses

sinamas committed:
add testrunner binary to .gitignore

sinamas committed:
libgambatte/gambatte: formatting/readability

sinamas committed:
libgambatte/interruptrequester: formatting/readability

sinamas committed:
libgambatte/mem: formatting/readability

sinamas committed:
libgambatte/memory: formatting/readability

sinamas committed:
libgambatte/minkeeper: formatting/readability

sinamas committed:
libgambatte/state_osd_elements: formatting/readability

sinamas committed:
libgambatte/video: formatting/readability

sinamas committed:
libgambatte/ppu: formatting/readability

sinamas committed:
common/chainresampler: formatting/readability

sinamas committed:
common/chainresampler: include order

sinamas committed:
trailing whitespace

sinamas committed:
libgambatte/bitmap_font: use size_t for text width

sinamas committed:
libgambatte/bitmap_font: use ptrdiff_t for pitch

sinamas committed:
libgambatte/state_osd_elements: use smarter array

sinamas committed:
common/adaptiveSleep: formatting/readability

sinamas committed:
common/rateest: formatting/readability

sinamas committed:
common/linint: include order

sinamas committed:
common/polyphasefir: include order

sinamas committed:
libgambatte/ppu: unindent

sinamas committed:
libgambatte/ppu: extract loadTileDataByte query functions

reduce duplication.

sinamas committed:
libgambatte/ppu: less magical lcdc constants

add named constants and lcdc query functions.
not everyone has lcdc specs in their heads.

sinamas committed:
libgambatte/ppu: fix broken indentation

probably introduced in 9f04bbc226d7cc3c0e64b90a4051bb2dd08eed41

sinamas committed:
libgambatte/ppu: formatting/readability

sinamas committed:
libgambatte/ppu: kill github horizontal scrollbar?

sinamas committed:
libgambatte/cartridge: allow mapping rom bank0 in more cases/please blargg

blargg seems to be of the opinion that mbc1 allows mapping bank0 via overflow
for carts with few rom banks if his tests are anything to go by. I have no way
of verifying this, but am changing it because it confuses people when his tests
loop forever after succeeding.

sinamas committed:
gambatte_qt/qglblitter: replace smart array with std::vector

sinamas committed:
vsync/frametime estimation improvements

use frametime samples below jitter limit as indication that we are not
blocking and slightly reduce current estimate to better support fifo triple
buffering implementations that implicitly avoid blocking on swapbuffers
when having a free buffer. intel linux opengl driver do this implicitly
when full screen for double buffered opengl, probably others do too. we
need to block to get usable estimates.

use a tighter sliding jitter window and allow slightly higher difference
from nominal rate because my quirky laptop thinks 60 Hz == 62.2 Hz.

sinamas committed:
gambatte_qt/framework: add common to dependpath

fixes no recompilation when touching common headers

sinamas committed:
gambatte_qt: simplify .pro files HEADER deps by using wildcards

also fixes some missing header deps.

sinamas committed:
gambatte_qt/framework: openbsd-*:LIBS += -lusbhid -lossaudio

sinamas committed:
gambatte_qt/ossengine: use /dev/audio on OpenBSD by default

requested by anthonyjbentley.

sinamas committed:
libgambatte/sound/channel4: formatting/readability/dead code

sinamas committed:
libgambatte/sound/channel3: formatting/readability

sinamas committed:
libgambatte/sound/channel2: formatting/readability

sinamas committed:
libgambatte/sound/channel1: formatting/readability

sinamas committed:
libgambatte/sound/dutyunit: formatting/readability

sinamas committed:
libgambatte/sound/length_counter: formatting/readability

sinamas committed:
libgambatte/sound: formatting/readability

sinamas committed:
libgambatte/cpu: formatting/readability

sinamas committed:
libgambatte/cpu: drop f/from_f macros

sinamas committed:
libgambatte/ppu: superflous mask

sinamas committed:
libgambatte/cpu: less magic HF2 constants

sinamas committed:
libgambatte/gambatte: formatting/readability

sinamas committed:
libgambatte: use ptrdiff_t to represent pitch

sinamas committed:
common/videolink: use ptrdiff_t to represent pitch

sinamas committed:
gambatte_sdl: use ptrdiff_t to represent pitch

sinamas committed:
gambatte_qt: use ptrdiff_t to represent pitch

sinamas committed:
gambatte_qt/framework/x11blitter: formatting/readability/cleanup

sinamas committed:
gambatte_qt/framework/xvblitter: formatting/readability/cleanup

sinamas committed:
gambatte_qt/gambattesource: drop use of volatile

the intention was probably to enforce ordering between stores to volatile
variables within the same thread, but that does not seem important here and
there's also no guarantee that they will be seen in that order from other
threads (which I knew so it must have been meant as an "increase probability"
thing), so it mostly just adds confusion and looks silly.

sinamas committed:
gambatte_qt/framework/qglblitter: clear thrice

more behind-your-back triple buffer semantics leaking through.

there does not seem to be any guarantee for how many buffers are actually
used with "double-buffered" opengl, but more than 3 seems unlikely.

contrary to popular belief avoiding clears actually matters in some relevant
scenarios, so let's just bump the number of clears to 3 and move on.

sinamas committed:
gambatte_qt/framework/qglblitter: formatting/readability/cleanup

sinamas committed:
qpainterblitter: always use image back buffer

get rid of union.
avoid touching back buffer on resize.
lock-free resize.

sinamas committed:
qpainterblitter: update new frontbuffer from backbuffer on resize

make an effort to keep image data on resize.

sinamas committed:
common: add smart_ptr deleter template parameter

sinamas committed:
xvblitter: use smart_ptrs to handle XFrees

sinamas committed:
x11blitter: use smart_ptrs to handle XFrees

sinamas committed:
x11blitter: hide class definition

get rid of Visual void *.

sinamas committed:
x11blitter: use image abstraction rather than subblitter abstraction

avoid touching back buffer on resize.
lock-free resizing.
more error checking.
use perror.

sinamas committed:
x11blitter: update new frontImage from backImage on resize if paused

make an effort to avoid losing paused picture on resize.

sinamas committed:
swscale: support srcpitch. less casting. gcc x86 tweaks.

sinamas committed:
qglblitter: only create subwidget when used

avoid stale state hanging around.
more proper use of initializeGL.

add some conservative use of makeCurrent again and hope that we do
not run into common driver bugs this time.

sinamas committed:
qglblitter: use QGLContext::getProcAddress rather than custom version

delete now unused x11getprocaddress.

sinamas committed:
qpainterblitter: hide class definition

sinamas committed:
xvblitter: hide class definition. improve shm error checking. cleanup.

use perror.
name constants.
drop some superfluous XSyncs.

sinamas committed:
qglblitter: hide class definition

sinamas committed:
defined_ptr: lower case defined_deleter

sinamas committed:
auto_vec: disallow element assignment. add reset method.

sinamas committed:
gambatte_sdl: fix print help exit success derp introduced in

44b6913a3ab7f2ecc1eca2abdbf0953d904b34d9

kill failed boolean. use EXIT_FAILURE.
initialize all members.

sinamas committed:
ringbuf: use array size in fill

sinamas committed:
gambatte_sdl/audiodata: cleanup. use smart_ptrs for sdl frees.

sinamas committed:
gambatte_sdl: blitter cleanup

sinamas committed:
gambatte_sdl: move event handling to a separate method

sinamas committed:
gambatte_sdl: replace printf with puts

sinamas committed:
gambatte_sdl: s/NULL/0

sinamas committed:
gambatte_sdl: move members to method scope

sinamas committed:
gambatte_sdl: use make_pair

sinamas committed:
gambatte_sdl: use value_type

sinamas committed:
gambatte_sdl: use std::for_each

sinamas committed:
gambatte_sdl: use sizeof array rather than repeat constant

sinamas committed:
gambatte_sdl: readability/formatting

sinamas committed:
gambatte_sdl/DescOption: use std::string

sinamas committed:
gambatte_sdl: generate desc string on demand now that it's possible

sinamas committed:
gambatte_sdl: name return value

sinamas committed:
gambatte_sdl: name condition

sinamas committed:
gambatte_sdl: reduce variable scope

sinamas committed:
gambatte_sdl: name constants

sinamas committed:
libgambatte/inputgetter: name enum

sinamas committed:
gambatte_sdl: encapsulate button mapping within inputoption

sinamas committed:
gambatte_sdl: use for_each to print option usage

sinamas committed:
gambatte_sdl: lower case constants/enums

sinamas committed:
gambatte_sdl: replace syncfunc with framewait object

sinamas committed:
gambatte_sdl: rename audiodata to audiosink

sinamas committed:
gambatte_sdl: enable failed field to be immutable

sinamas committed:
gambatte_sdl: define dtor to allow forward declaration

sinamas committed:
gambatte_sdl/str_to_sdlkey: use make_pair

sinamas committed:
gambatte_sdl: lose empty checks

sinamas committed:
gambatte_sdl: unsued variable warnings

sinamas committed:
gambatte_sdl/parser: formatting/readability

sinamas committed:
gambatte_sdl/parser: make parse methods const

sinamas committed:
gambatte_sdl/parser: use make_pair

sinamas committed:
libgambatte/video: more query functions

sinamas committed:
libgambatte/video: formatting/readability

sinamas committed:
libgambatte/video: move m0irq to its own header

sinamas committed:
libgambatte: pass intreq by ref

sinamas committed:
libgambatte/ppu: const placement consistency

sinamas committed:
libgambatte/ppu: public/private order

sinamas committed:
libgambatte/ppu: lower case constants/enums

sinamas committed:
libgambatte/ppu: avoid depending on magic win_draw_state values

sinamas committed:
libgambatte/ppu: broken indent

sinamas committed:
libgambatte/sound: drop use of WORDS_BIGENDIAN define

sinamas committed:
gambatte_sdl: avoid clang struct vs class warning

sinamas committed:
common: drop use of WORDS_BIGENDIAN

sinamas committed:
libgambatte: less magic lcd constants

sinamas committed:
libgambatte: cosmetics/consistency

sinamas committed:
libgambatte: shave off an instruction or two on ff loads/stores

gcc arm/x86.

sinamas committed:
libgambatte: clarify reasoning behind interface type usage

sinamas committed:
gambatte_sdl/blitter: cosmetics/consistency[/ocd]

sinamas committed:
gambatte_qt: non-framework code formatting/readability/consistency

sinamas committed:
gambatte_qt/framework/samplebuffer: ensure 32-bit alignment

in practice this should already be fine, but it's nice to comply with
c++98 which only guarantees alignment beyond the type allocated for char
arrays. just use quint32 here.

sinamas committed:
gambatte_qt/framework: add framework/include to dependpath

sinamas committed:
gambatte_qt: pass by ref

sinamas committed:
gambatte_qt: fix -m64 build breakage

sinamas committed:
gambatte_qt: cleanup [/waytoomuch]

sinamas committed:
gambatte_qt: less misleading indent?

sinamas committed:
libgambatte: fix lazy rgb32 type [/wtf][/abibreak]

sinamas committed:
gambatte_qt: fix win32 build breakage

sinamas committed:
common: cleanup

sinamas committed:
common/videolink: cleanup/minor tweaks

use ptrdiff prevline offset in hqx filters to facilitate use of lea.

use endptr rathern than loop ctr in rgb2yuv to avoid gcc/x86 rmw
spill.

avoid reloading rhs input pixel after storing lhs output in rgb2yuv
(no restrict keyword in c++98).

fix hypothetical component overflow in hq2x blend9.

sinamas committed:
common/resample: cleanup

rename polyphaseconvoluter to polyphasefir.

fix broken upsampler zero stuffing (was never tested, surprisingly the
surrounding code appears to work fine).

sinamas committed:
common: fix rgb32ToUyvy brain fart

sinamas committed:
common/chainresampler: hide more implementation

sinamas committed:
gambatte_qt: add libgambatte/include to DEPENDPATH

sinamas committed:
more ideomatic types [/apibreak]

sinamas committed:
libgambatte: broken indent

sinamas committed:
common/chainresampler: less template bloat

sinamas committed:
libgambatte: s/drawn/completed etc.

sinamas committed:
libgambatte: drop useless term and emphasis

sinamas committed:
update description

sinamas committed:
attempt to insert revision number automatically

Qt solution needs some work to trigger rebuild.

sinamas committed:
gambatte_qt/framework: avoid promoting -1 to size_t

sinamas committed:
reword description

bentley committed:
Use arrays instead of dicts. Fixes build on OpenBSD.

bentley committed:
testrunner: respect environment-defined CXX (like other gambatte parts do).

sinamas committed:
libgambatte: fix lfsr reg init value

all bits should be high, not just the lower 8.

most likely this is a remnant from the initial implementation
based on GBSOUND.txt which has what is most likely a brain fart
0xFF value for this considering the rest of the description
is inconsistent whith that.

belmont's revenge (which has some great tracks) is affected by
this, so this one is a pretty big deal to me. konami collection
vol 4 has a pretty sweet color version of it which I recommend
checking out.

sinamas committed:
Merge pull request #1 from bentley/sconsdict

Use arrays instead of dicts. Fixes build on OpenBSD.

sinamas committed:
Merge pull request #2 from bentley/cxxtest

testrunner: respect environment-defined CXX (like other gambatte parts do).

sinamas committed:
gambatte_qt/framework: fix broken js event filtering

do continue while false==true ...

sinamas committed:
commen/rateest: somewhat faster convergence

utilize intermediate measurements to accumulate a least squares solution.

sinamas committed:
gambatte_qt/framework: slightly faster frametime estimation convergence

utilize intermediate measurements to accumulate a least squares solution.

sinamas committed:
hwtests: less cryptic

sinamas committed:
hwtests: s/.code/.text/

sinamas committed:
hwtests: less cryptic

sinamas committed:
hwtests: copy-paste fixup

sinamas committed:
hwtests: cleanup

sinamas committed:
hwtests: cleanup

sinamas committed:
gambatte_qt/framework: default to direct3d on win32

sinamas committed:
gambatte_qt/framework: mark default audio engine as recommended

sinamas committed:
gambatte_qt/framework: missing include

fixes build on recent os x toolchain

sinamas committed:
new clang warnings

bentley committed:
Print more informative error messages when using MBC6 or MBC7.

sinamas committed:
libgambatte: do not default initialize etc.

sinamas committed:
libgambatte: audio findings

sinamas committed:
testrunner: use stderr

sinamas committed:
fix description.

sinamas committed:
reword slightly.

sinamas committed:
libgambatte: joypad irq findings.

sinamas committed:
s/simple/simplistic/.

sinamas committed:
Merge pull request #3 from bentley/master

Print more informative error messages when using MBC6 or MBC7.

sinamas committed:
rename tests to suggest init state dependency.

sinamas committed:
libgambatte: fix (and verify) strange key1 wtf.

not sure what this originated from.

sinamas committed:
rename test directory.

sinamas committed:
libgambatte: fixup remaining old-style headers.

sinamas committed:
libgambatte: make mbc3 behave more like mbc1.

map rom bank 0 to bank 1.

fixes pokemon red/blue name entry glitch that has gone unnoticed for
years.

found by anthonyjbentley.

sinamas committed:
change description again. probably not the last time.

sinamas committed:
simplify description.

sinamas committed:
test: add a textual description of jpad irq observations.

this would seem worthwile with the relevant tests lacking a
specification of expected output (for now at least).

sinamas committed:
description of the day.

sinamas committed:
qualify.

sinamas committed:
simpler.

sinamas committed:
adjust copyright header.

bentley committed:
Update FSF address.

sinamas committed:
libgambatte: push return address after pc mod.

unsurprising that the call target is fetched prior to
pushing the return address (sufficiently so to verify
this based on wtf-factor alone, really). more surprising
that the 4 cycle delay that accompanies pc mods also
precedes it.

"woody woodpecker racing" barely avoids an oam dma bus
conflict due to this fact.

bentley committed:
Add a Makefile for building libgambatte and gambatte_sdl.

hasufell commented on Makefile:
```diff
@@ -0,0 +1,114 @@
+
+SDL_LFLAGS != pkg-config --libs sdl
+$(SDL): $(SDL_OBJECTS) $(LIB)
+	$(CXX) $(CFLAGS) -o $@ -pthread $(SDL_OBJECTS) $(LIB) -lz \
```
This should be CXXFLAGS, also LDFLAGS are missing.

hasufell commented on Makefile:
```diff
@@ -0,0 +1,114 @@
+
+all: $(SDL)
+
+SDL_LFLAGS != pkg-config --libs sdl
```
I'd declare pkg-config in a variable PKG_CONFIG, that makes things easier for a lot of use cases (e.g. cross compiling with a wrapper script).

hasufell commented on Makefile:
```diff
@@ -0,0 +1,114 @@
+
+$(LIB): $(LIB_OBJECTS)
+	$(AR) $(ARFLAGS) $@ $(LIB_OBJECTS)
+	ranlib $@
```
ranlib should also be declared in a variable

hasufell commented on Makefile:
```diff
@@ -0,0 +1,114 @@
+	$(CC) $(CFLAGS) -c $< -o $*.o
+.cpp.o:
+	$(CXX) -Ilibgambatte/src -Ilibgambatte/include -Igambatte_sdl/src \
+		-Icommon $(CFLAGS) -c $< -o $*.o \
```
should be CXXFLAGS

hasufell commented on Makefile:
```diff
@@ -0,0 +1,114 @@
+SDL = gambatte_sdl/gambatte_sdl
+TEST = test/testrunner
+
+PY = python
```
Some distros use the environment variable PYTHON to handle things. I also find the name better.
```
PYTHON ?= python
```
for example, but both ways work ofc.

hasufell commented:
It seems both install and uninstall rules are missing. Would be good to have them too.

bentley commented:
install is not a bad thing, but I didn’t bother with it because 1) there isn’t one already, and 2) it is harder to get right (handle DESTDIR? handle PREFIX? install library? install headers?…)

hasufell commented on Makefile:
```diff
@@ -0,0 +1,115 @@
+PKG_CONFIG = pkg-config
+PKGCONFIG_CFLAGS != $(PKG_CONFIG) --cflags sdl libpng
+.c.o:
+	$(CC) $(CXXFLAGS) -c $< -o $*.o
```
this should be $(CFLAGS) afais, because of CC (the other CXXFLAGS look correct though)

hasufell commented on Makefile:
```diff
@@ -0,0 +1,115 @@
+SDL_LFLAGS != $(PKG_CONFIG) --libs sdl
+$(SDL): $(SDL_OBJECTS) $(LIB)
+	$(CXX) $(CXXFLAGS) -o $@ -pthread $(SDL_OBJECTS) $(LIB) -lz \
+		$(SDL_LFLAGS) $(LDFLAGS)
```
LDFLAGS (like ```-Wl,-O1```) only apply to libraries that follow them, so it should be earlier in the linker line... directly after CXXFLAG or after the '-o' target

hasufell commented on Makefile:
```diff
@@ -0,0 +1,115 @@
+	$(AR) $(ARFLAGS) $@ $(LIB_OBJECTS)
+	$(RANLIB) $@
+
+PKG_CONFIG = pkg-config
```
declared too late... it will make line 74 fail silently

hasufell commented:
I'll prepare a PR to your repo

hasufell commented on Makefile:
```diff
@@ -0,0 +1,146 @@
+
+PKGCONFIG_CFLAGS != $(PKG_CONFIG) --cflags sdl libpng zlib
+.c.o:
+	$(CC) $(CXXFLAGS) -c $< -o $*.o
```
oops, still missed this one :P

hasufell commented:
any progress here?

sinamas committed:
Merge pull request #5 from bentley/fsfaddr

Update FSF address.

sinamas committed:
libgambatte: avoid bools in save state struct.

ensure boolean conversion on assignment.

sinamas committed:
common/resample: move include from .h to .cpp

hasufell commented:
we could probably squash the commits, right @bentley

sinamas committed:
hwtests: indicate cgb revision tested.

bentley commented:
Squashed and rebased the commits.

sergiobenrocha2 commented:
Didn't work on debian based, sdl headers needs to be like this

#include <SDL/SDL.h>

In lot of files, and undefined references (missing -lSDL and others?)

bentley commented:
Works for me on Jessie. Can you be more specific?

sergiobenrocha2 commented:
```
g++   -o gambatte_sdl/gambatte_sdl -pthread gambatte_sdl/src/audiosink.o gambatte_sdl/src/blitterwrapper.o gambatte_sdl/src/gambatte_sdl.o gambatte_sdl/src/parser.o gambatte_sdl/src/sdlblitter.o gambatte_sdl/src/str_to_sdlkey.o gambatte_sdl/src/usec.o common/adaptivesleep.o common/resample/src/chainresampler.o common/resample/src/i0.o common/resample/src/kaiser50sinc.o common/resample/src/kaiser70sinc.o common/resample/src/makesinckernel.o common/resample/src/resamplerinfo.o common/resample/src/u48div.o common/rateest.o common/skipsched.o common/videolink/rgb32conv.o common/videolink/vfilterinfo.o common/videolink/vfilters/catrom2x.o common/videolink/vfilters/catrom3x.o common/videolink/vfilters/kreed2xsai.o common/videolink/vfilters/maxsthq2x.o common/videolink/vfilters/maxsthq3x.o libgambatte/libgambatte.a \
		 
gambatte_sdl/src/audiosink.o: In function `(anonymous namespace)::openAudio(long, unsigned long, void (*)(void*, unsigned char*, int), void*)':
audiosink.cpp:(.text+0xb9): undefined reference to `SDL_OpenAudio'
audiosink.cpp:(.text+0xc5): undefined reference to `SDL_GetError'
gambatte_sdl/src/audiosink.o: In function `(anonymous namespace)::LockGuard::LockGuard(SDL_mutex*)':
audiosink.cpp:(.text+0x120): undefined reference to `SDL_mutexP'
gambatte_sdl/src/audiosink.o: In function `(anonymous namespace)::LockGuard::LockDeleter::del(SDL_mutex*)':
audiosink.cpp:(.text+0x15b): undefined reference to `SDL_mutexV'
gambatte_sdl/src/audiosink.o: In function `AudioSink::AudioSink(long, int, int)':
audiosink.cpp:(.text+0x1f1): undefined reference to `SDL_CreateMutex'
audiosink.cpp:(.text+0x20c): undefined reference to `SDL_CreateCond'
gambatte_sdl/src/audiosink.o: In function `AudioSink::~AudioSink()':
audiosink.cpp:(.text+0x2f7): undefined reference to `SDL_PauseAudio'
audiosink.cpp:(.text+0x2fc): undefined reference to `SDL_CloseAudio'
gambatte_sdl/src/audiosink.o: In function `AudioSink::write(short const*, unsigned long)':
audiosink.cpp:(.text+0x4f4): undefined reference to `SDL_CondWait'
gambatte_sdl/src/audiosink.o: In function `AudioSink::read(unsigned char*, unsigned long)':
audiosink.cpp:(.text+0x654): undefined reference to `SDL_CondSignal'
gambatte_sdl/src/audiosink.o: In function `AudioSink::SdlDeleter::del(SDL_mutex*)':
audiosink.cpp:(.text._ZN9AudioSink10SdlDeleter3delEP9SDL_mutex[_ZN9AudioSink10SdlDeleter3delEP9SDL_mutex]+0x14): undefined reference to `SDL_DestroyMutex'
gambatte_sdl/src/audiosink.o: In function `AudioSink::SdlDeleter::del(SDL_cond*)':
audiosink.cpp:(.text._ZN9AudioSink10SdlDeleter3delEP8SDL_cond[_ZN9AudioSink10SdlDeleter3delEP8SDL_cond]+0x14): undefined reference to `SDL_DestroyCond'
gambatte_sdl/src/gambatte_sdl.o: In function `(anonymous namespace)::SdlIniter::SdlIniter()':
gambatte_sdl.cpp:(.text+0x110a): undefined reference to `SDL_Init'
gambatte_sdl.cpp:(.text+0x1125): undefined reference to `SDL_GetError'
gambatte_sdl/src/gambatte_sdl.o: In function `(anonymous namespace)::SdlIniter::~SdlIniter()':
gambatte_sdl.cpp:(.text+0x1155): undefined reference to `SDL_Quit'
gambatte_sdl/src/gambatte_sdl.o: In function `(anonymous namespace)::JsOpen::~JsOpen()':
gambatte_sdl.cpp:(.text+0x1197): undefined reference to `SDL_JoystickClose'
gambatte_sdl/src/gambatte_sdl.o: In function `(anonymous namespace)::GambatteSdl::exec(int, char const* const*)':
gambatte_sdl.cpp:(.text+0x206d): undefined reference to `SDL_InitSubSystem'
gambatte_sdl.cpp:(.text+0x2086): undefined reference to `SDL_GetError'
gambatte_sdl.cpp:(.text+0x20ef): undefined reference to `SDL_JoystickEventState'
gambatte_sdl.cpp:(.text+0x2159): undefined reference to `SDL_ShowCursor'
gambatte_sdl.cpp:(.text+0x2168): undefined reference to `SDL_WM_SetCaption'
gambatte_sdl/src/gambatte_sdl.o: In function `(anonymous namespace)::GambatteSdl::handleEvents(BlitterWrapper&)':
gambatte_sdl.cpp:(.text+0x2abd): undefined reference to `SDL_PollEvent'
gambatte_sdl/src/gambatte_sdl.o: In function `(anonymous namespace)::GambatteSdl::run(long, int, int, ResamplerInfo const&, BlitterWrapper&)':
gambatte_sdl.cpp:(.text+0x2c12): undefined reference to `SDL_GetKeyState'
gambatte_sdl.cpp:(.text+0x2c35): undefined reference to `SDL_PauseAudio'
gambatte_sdl/src/gambatte_sdl.o: In function `(anonymous namespace)::JsOpen::JsOpen<std::_Rb_tree_const_iterator<unsigned char> >(std::_Rb_tree_const_iterator<unsigned char>, std::_Rb_tree_const_iterator<unsigned char>)':
gambatte_sdl.cpp:(.text+0x348a): undefined reference to `SDL_JoystickOpen'
gambatte_sdl/src/sdlblitter.o: In function `SdlBlitter::SdlBlitter(unsigned int, unsigned int, int, bool, bool)':
sdlblitter.cpp:(.text+0x35): undefined reference to `SDL_GetVideoInfo'
sdlblitter.cpp:(.text+0x68): undefined reference to `SDL_SetVideoMode'
sdlblitter.cpp:(.text+0xcb): undefined reference to `SDL_CreateRGBSurface'
sdlblitter.cpp:(.text+0x11b): undefined reference to `SDL_CreateYUVOverlay'
sdlblitter.cpp:(.text+0x161): undefined reference to `SDL_LockYUVOverlay'
gambatte_sdl/src/sdlblitter.o: In function `SdlBlitter::present()':
sdlblitter.cpp:(.text+0x41b): undefined reference to `SDL_UnlockYUVOverlay'
sdlblitter.cpp:(.text+0x43a): undefined reference to `SDL_DisplayYUVOverlay'
sdlblitter.cpp:(.text+0x452): undefined reference to `SDL_LockYUVOverlay'
sdlblitter.cpp:(.text+0x48a): undefined reference to `SDL_UpdateRect'
gambatte_sdl/src/sdlblitter.o: In function `SdlBlitter::toggleFullScreen()':
sdlblitter.cpp:(.text+0x4e1): undefined reference to `SDL_SetVideoMode'
gambatte_sdl/src/sdlblitter.o: In function `SdlBlitter::SurfaceDeleter::del(SDL_Surface*)':
sdlblitter.cpp:(.text._ZN10SdlBlitter14SurfaceDeleter3delEP11SDL_Surface[_ZN10SdlBlitter14SurfaceDeleter3delEP11SDL_Surface]+0x14): undefined reference to `SDL_FreeSurface'
gambatte_sdl/src/sdlblitter.o: In function `SdlBlitter::SurfaceDeleter::del(SDL_Overlay*)':
sdlblitter.cpp:(.text._ZN10SdlBlitter14SurfaceDeleter3delEP11SDL_Overlay[_ZN10SdlBlitter14SurfaceDeleter3delEP11SDL_Overlay]+0x1b): undefined reference to `SDL_UnlockYUVOverlay'
sdlblitter.cpp:(.text._ZN10SdlBlitter14SurfaceDeleter3delEP11SDL_Overlay[_ZN10SdlBlitter14SurfaceDeleter3delEP11SDL_Overlay]+0x27): undefined reference to `SDL_FreeYUVOverlay'
gambatte_sdl/src/usec.o: In function `getusecs()':
usec.cpp:(.text+0x5): undefined reference to `SDL_GetTicks'
gambatte_sdl/src/usec.o: In function `usecsleep(unsigned long)':
usec.cpp:(.text+0x45): undefined reference to `SDL_Delay'
libgambatte/libgambatte.a(file_zip.o): In function `(anonymous namespace)::GzFile::GzFile(char const*)':
file_zip.cpp:(.text+0x490): undefined reference to `gzopen'
file_zip.cpp:(.text+0x503): undefined reference to `gzread'
libgambatte/libgambatte.a(file_zip.o): In function `(anonymous namespace)::GzFile::rewind()':
file_zip.cpp:(.text+0x641): undefined reference to `gzrewind'
libgambatte/libgambatte.a(file_zip.o): In function `(anonymous namespace)::GzFile::read(char*, unsigned long)':
file_zip.cpp:(.text+0x6b4): undefined reference to `gzread'
libgambatte/libgambatte.a(file_zip.o): In function `(anonymous namespace)::GzFile::close()':
file_zip.cpp:(.text+0x717): undefined reference to `gzclose'
libgambatte/libgambatte.a(unzip.o): In function `unzOpenCurrentFile3':
unzip.c:(.text+0x1fb8): undefined reference to `inflateInit2_'
unzip.c:(.text+0x2065): undefined reference to `get_crc_table'
libgambatte/libgambatte.a(unzip.o): In function `unzReadCurrentFile':
unzip.c:(.text+0x265f): undefined reference to `crc32'
unzip.c:(.text+0x272d): undefined reference to `inflate'
unzip.c:(.text+0x2788): undefined reference to `crc32'
libgambatte/libgambatte.a(unzip.o): In function `unzCloseCurrentFile':
unzip.c:(.text+0x2aa4): undefined reference to `inflateEnd'
collect2: error: ld returned 1 exit status
make: *** [gambatte_sdl/gambatte_sdl] Error 1

```

sergiobenrocha2 commented:
*Ubuntu 14.04

bentley commented:
Did you try installing pkg-config?

sergiobenrocha2 commented:
already installed

bentley commented:
And what do `pkg-config --libs sdl` and `pkg-config --libs zlib` give? It looks very much like your pkg-config is not working properly.

sergiobenrocha2 commented:
pkg-config --libs sdl --> -lSDL
pkg-config --libs zlib --> -lz

hasufell commented:
Try this patch

```
--- a/Makefile
+++ b/Makefile
@@ -87,8 +87,8 @@ TEST_OBJECTS = \

 all: $(SDL_TARGET)

-SDL_LFLAGS != $(PKG_CONFIG) --libs sdl
-ZLIB_LFLAGS != $(PKG_CONFIG) --libs zlib
+SDL_LFLAGS = $(shell $(PKG_CONFIG) --libs sdl)
+ZLIB_LFLAGS = $(shell $(PKG_CONFIG) --libs zlib)
 $(SDL_TARGET): $(SDL_OBJECTS) $(LIB)
        $(CXX) $(CXXFLAGS) $(LDFLAGS) -o $@ -pthread $(SDL_OBJECTS) $(LIB) \
                $(ZLIB_LFLAGS) $(SDL_LFLAGS)
@@ -97,7 +97,7 @@ $(LIB): $(LIB_OBJECTS)
        $(AR) $(ARFLAGS) $@ $(LIB_OBJECTS)
        $(RANLIB) $@

-PKGCONFIG_CFLAGS != $(PKG_CONFIG) --cflags sdl libpng zlib
+PKGCONFIG_CFLAGS = $(shell $(PKG_CONFIG) --cflags sdl libpng zlib)
 .c.o:
        $(CC) $(CFLAGS) -c $< -o $*.o
 .cpp.o:
@@ -120,7 +120,7 @@ test: $(TEST)
                test/hwtests/*/*/*/*.asm
        $(TEST) $(TEST_GBS)

-PNG_LFLAGS != $(PKG_CONFIG) --libs libpng
+PNG_LFLAGS = $(shell $(PKG_CONFIG) --libs libpng)
 $(TEST): $(TEST_OBJECTS) $(LIB)
        $(CXX) $(CXXFLAGS) -o $@ $(TEST_OBJECTS) $(LIB) \
                $(PNG_LFLAGS) $(ZLIB_LFLAGS)
```

bentley commented:
Ah, that’s probably it. Unfortunately, neither `$(shell foo)` nor `!=` are portable. But `!=` is supported by GNU Make 4.0 and greater as well as BSD Make and there have been attempts to get it into POSIX, while `$(shell foo)` is only supported by GNU Make.

Once Ubuntu updates their GNU Make package, the problem will go away on its own.

sergiobenrocha2 commented:
"Once Ubuntu updates their GNU Make package, the problem will go away on its own."

Problem is ubuntu 14.04 is LTS... there are more 4 years of support.

sergiobenrocha2 commented:
The patch worked!

sergiobenrocha2 commented:
Nice, I was getting hardening-no-relro lintian error in sdl binary with scons, now with makefile it's fine.

Is it possible to merge this?

hasufell commented:
> GNU Make 4.0 and greater as well as BSD Make and there have been attempts to get it into POSIX

just make BSD.Makefile and GNU.Makefile and call the correct one from within Makefile depending on some hacky checks... if these checks don't work, people can do `make -f BSD.Makefile` manually

bentley committed:
Add a thin wrapper makefile for GNU Make.

GNU Make 3.x doesn't support the "!=" syntax for shell assignment,
although 4.x does, BSD Makes do, and it has been submitted to POSIX
for possible inclusion in the standard. Unfortunately, GNU Make 3.x
is still in common use, so provide an alternative makefile (GNU Make
reads "GNUmakefile" instead of "Makefile" by default) that sets the
variables and then reuses the rest of the default Makefile.

bentley commented:
> just make BSD.Makefile and GNU.Makefile and call the correct one from within Makefile depending on some hacky checks...

That’s exactly the opposite of my intention here… the point of switching to make is to be simple, fast, and portable. Duplicating makefiles and adding hackish checks over a single feature missing in a nearly two‐year‐old GNU Make is something I emphatically don’t want.

Here’s an alternative way without any hackish checks… GNU Make will read GNUmakefile by default, so we can set the variables in there with `$(shell foo)` and then include the regular Makefile. This will catch GNU Make 4.x as well, but that’s harmless. Other makes are unaffected.

sergiobenrocha2 commented:
Any news?

bentley commented:
I just tweaked the compiler defaults bit. Other than that, I’ve been using this makefile unmodified for several months.

andychase commented:
I like the simplicity of this but I'm not sure how well this would work for Windows builds (with the native VC compiler). Scons does a few conversions under the hood to make it work.

This might work with Mysys/Mingw but I don't know how to set that up.

hasufell commented:
> I like the simplicity of this but I'm not sure how well this would work for Windows builds

This is an addition, not a replacement of SCons.

andychase commented:
The title of this issue says "switch". Also it usually not a good idea to have duplicated logic unless there is a really good reason.

hasufell commented:
> Also it usually not a good idea to have duplicated logic unless there is a really good reason.

I've been a distro package for more than 4 years and I can assure you that everyone hates SCons (especially source distros) since it needs various hacks to work correctly. And with "correctly" I don't just mean "it compiles". This Makefile makes things easier for that audience, so it's not really "duplicated logic".

Besides that, compilation time is reduced. Please see the initial post of this PR.

HunterZ commented:
Banging my head on SCons right now because I am trying to cross-compile for an arm platform using a custom cross-compiler toolchain. SCons sucks at this use case.

Mark me down as another proponent of changing to a build system that doesn't suck. What's wrong with CMake? Gambatte is not doing anything fancy enough to make SCons is a huge advantage here...

DMinsky commented:
I have had zero success in building this using _build_sdl.sh_ on Fedora 27 with following error:

`scons: Reading SConscript files ...
Checking for C library z... yes
TypeError: must be str, not bytes:
  File "/home/dminsky/Projects/gambatte/gambatte_sdl/SConstruct", line 51:
    version_str_def = [ 'GAMBATTE_SDL_VERSION_STR', r'\"r' + git_revno + r'\"' ]
`

But with _make_ it works like a charm.

filpAM commented:
...Why is this still stuck on a limbo?

I agree having a Makefile would be really useful for gcc-toolchain users but I would rather not "Switch"(as the title suggest) but have it as an additional building tool, since sinamas and some other contribuitors still use scons.

yoyz commented:
Really put a makefile next to the scons build things.
I really don't like to see scons, cmake stuff everywhere.
It's really cool when this tool work, but could be a nightmare when you try just to build the software and scons/cmake break.
And tell in the README that scons is the "preffered" way of building the software.
So you won't bother too much with support request.
People like me preffered to work with Makefile in the end.

HunterZ commented:
Normally I'd prefer CMake because it's actively maintained and if done
right requires a minimum of fiddling, and is relatively easy to maintain.

Unfortunately CMake appears to require a lot of spoon-feeding when doing
cross compilation, which is probably a major use case for Gambatte.

GNU autotools supports cross compilation fairly well, but I strongly
dislike everything else about it.

I don't know much about scons.

On Tue, Apr 9, 2019 at 1:27 AM yoyz <notifications@github.com> wrote:

> Really put a makefile next to the scons build things.
> I really don't like to see scons, cmake stuff everywhere.
> It's really cool when this tool work, but could be a nightmare when you
> try just to build the software and scons/cmake break.
> And tell in the README that scons is the "preffered" way of building the
> software.
> So you won't bother too much with support request.
> People like me preffered to work with Makefile in the end.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sinamas/gambatte/pull/6#issuecomment-481154470>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGpi9PL4uu6_ueRfptXnuBpWtiKgmaoks5vfE8DgaJpZM4Dj5HB>
> .
>
